### PR TITLE
JAMES-3715 Add TCP transport level tests for IMAP IDLE

### PR DESCRIPTION
**Pull request by chibenwa**
In the face of Netty4 migration, pipeline modifications (which IDLE does)
deserves to be seriously tested.

### CODE REVIEW

File: server/protocols/protocols-imap4/src/test/java/org/apache/james/imapserver/netty/IMAPServerTest.java [hunk 1]
```diff
@@ -970,4 +977,106 @@ void searchingASingleUTF8CriterionShouldComplete() throws Exception {
...
+        }
+
+        @AfterEach
+        void tearDown() {
+            imapServer.destroy();
+        }
+
+        @Test
+        void idleShouldSendInitialContinuation() throws Exception {
+            SocketChannel server = SocketChannel.open();
```
Comment (quantranhong1999) [inline server/protocols/protocols-imap4/src/test/java/org/apache/james/imapserver/netty/IMAPServerTest.java:1007]: This is more like a client than a server? BTW should we put this SocketChannel.open() into BeforeEach and close it in AfterEach?

Comment (chibenwa) [reply server/protocols/protocols-imap4/src/test/java/org/apache/james/imapserver/netty/IMAPServerTest.java:1007]: > This is more like a client than a server?

Yes we are opening a TCP client connection to test the server.

> BTW should we put this SocketChannel.open() into BeforeEach and close it in AfterEach?

:+1:

File: server/protocols/protocols-imap4/src/test/java/org/apache/james/imapserver/netty/IMAPServerTest.java [hunk 2]
```diff
@@ -1034,11 +1038,73 @@ void idleShouldBeInterruptible() throws Exception {
...
+
+            server.write(ByteBuffer.wrap(("a3 IDLE\r\nDONE\r\n").getBytes(StandardCharsets.UTF_8)));
+
+            Awaitility.await().atMost(Duration.ofSeconds(2)).untilAsserted(() ->
+                assertThat(readStringUntil(server, s -> s.contains("a3 OK IDLE completed.")))
+                    .isNotNull());
         }
 
         @Test
+        void idleShouldResponsesShouldBeOrdered() throws Exception {
```
Comment (Arsnael) [inline server/protocols/protocols-imap4/src/test/java/org/apache/james/imapserver/netty/IMAPServerTest.java:1061]:
```suggestion
        void idleResponsesShouldBeOrdered() throws Exception {
```

File: server/protocols/protocols-imap4/src/test/java/org/apache/james/imapserver/netty/IMAPServerTest.java [hunk 3]
```diff
@@ -995,134 +996,115 @@ void beforeEach() throws Exception {
...
+            clientConnection = SocketChannel.open();
+            clientConnection.connect(new InetSocketAddress(LOCALHOST_IP, port));
+            readBytes(clientConnection);
         }
 
         @AfterEach
-        void tearDown() {
+        void tearDown() throws Exception {
             imapServer.destroy();
+            clientConnection.close();
```
Comment (Arsnael) [inline server/protocols/protocols-imap4/src/test/java/org/apache/james/imapserver/netty/IMAPServerTest.java:1008]: Shouldn't we close the client connection before destroying the server?

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Pull request merged**